src/lib/unlang/xlat_eval.c

```diff
@@ -2216,3 +2216,78 @@ bool xlat_async_required(xlat_exp_t const *xlat)
 
 	return false;
 }
+
+
+/** Parse a list of value boxes representing a URI
+ *
+ * Reads a URI from a list of value boxes and parses it according to the
+ * definition in uri_parts.  Tainted values, where allowed, are escaped
+ * using the function specified for the uri part.
+ *
+ * @param request	currently being processed
+ * @param uri		to parse
+ * @param uri_parts	definition of URI structure
+ * @param uctx		to pass to escaping function
+ * @return
+ * 	- 0 on success
+ * 	- -1 on failure
+ */
+int xlat_parse_uri(request_t *request, fr_value_box_list_t *uri, xlat_uri_part_t const *uri_parts, void *uctx)
+{
+	fr_value_box_t		*uri_vb = NULL;
+	xlat_uri_part_t const	*uri_part;
+	fr_sbuff_t		sbuff;
+	char const		*p;
+
+	uri_part = uri_parts;
+
+	while ((uri_vb = fr_dlist_next(uri, uri_vb))){
+		if (uri_vb->tainted && !uri_part->tainted_allowed) {
+			REDEBUG("Tainted value not allowed for %s", uri_part->name);
+			return -1;
+		}
+
+		/*
+		 *	Tainted boxes can only belong to a single part of the URI
+		 */
+		if (uri_vb->tainted) {
+			if ((uri_part->func) && (uri_part->func(request, uri_vb, uctx) < 0)) {
+				REDEBUG("Unable to escape tainted input %pV", uri_vb);
+				return -1;
+			}
+			continue;
+		}
+
+		/*
+		 *	This URI part has no term chars - so no need to look for them
+		 */
+		if (!uri_part->term_chars) continue;
+
+		/*
+		 *	Zero length box - no terminators here
+		 */
+		if (uri_vb->length == 0) continue;
+
+		/*
+		 *	Look for URI part terminator
+		 */
+		fr_sbuff_init(&sbuff, uri_vb->vb_strvalue, uri_vb->length);
+
+		do {
+			if ((p = strchr(uri_part->term_chars, *fr_sbuff_current(&sbuff)))) {
```

Comment:
There's sbuff functions to handle searcing for the next char

Reply:
```
	fr_sbuff_term_t const 		bareword_terminals =
					FR_SBUFF_TERMS(
						L("\t"),
						L("\n"),
						L(" "),
						L("%"),
						L("&"),
						L(")"),
						L("+"),
						L("-"),
						L("/"),
						L("^"),
						L("|")
					);
```

Reply:
```size_t	fr_sbuff_adv_until(fr_sbuff_t *sbuff, size_t len, fr_sbuff_term_t const *tt, char escape_chr);
```

Reply:
Then the sbuff will be left at the first character of the termination substring.  The terminal sequences do need to be ordered in the array though.